src/sentry/ratelimits/sliding_windows.py

```diff
@@ -0,0 +1,209 @@
+"""
+The ratelimiter used by the metrics string indexer to rate-limit DB writes.
+
+As opposed to the API rate limiter, a limit in the "sliding window" rate
+limiter such as "10 requests / minute" does not reset to 0 every minute.
+Instead each window can be configured with a "granularity" setting such that
+the window gradually resets in steps of `granularity` seconds.
+
+Additionally this rate-limiter is not coupled to per-project/organization
+scopes, and can apply multiple sliding windows at once. On the flipside it is
+not strongly consistent and depending on usage it is very easy to over-spend
+quota, as checking quota and spending quota are two separate steps.
+"""
+
+from dataclasses import dataclass
+from time import time
+from typing import Any, Iterator, Optional, Sequence, Tuple
+
+from sentry.exceptions import InvalidConfiguration
+from sentry.utils import redis
+from sentry.utils.services import Service
+
+
+@dataclass(frozen=True)
+class Quota:
+    # The number of seconds to apply the limit to.
+    window: int
+
+    # A number between 1 and `window`. Since `window` is a sliding window,
+    # configure what the granularity of that window is.
+    #
+    # If this is equal to `window`, the quota resets to 0 every `window`
+    # seconds. If this is a very small number, the window slides "more
+    # smoothly" at the expense of having much more redis keys.
+    #
+    # The number of redis keys required to enforce a quota is `window /
+    # granularity`.
+    granularity: int
+
+    #: How many units are allowed within the given window.
+    limit: int
+
+    # Override the prefix given by RequestedQuota such that one can implement
+    # global limits + per-organization limits. The GrantedQuota will still only
+    # contain the prefix of the RequestedQuota
+    prefix_override: Optional[str] = None
+
+    def iter_granules(self, request_timestamp: int) -> Iterator[int]:
+        assert self.window % self.granularity == 0
+        for granule_i in range(int(self.window / self.granularity)):
+            yield int(request_timestamp / self.granularity) - granule_i * self.granularity
```

Comment:
kind of a generic python question -- is `int(num/div)` preferable to `num // div` for some reason?

Reply:
nope i forgot how to write python